### PR TITLE
Phase 1d: Shaka Lab Node on macOS

### PR DESCRIPTION
**Pull request by joeyparrish**

### CODE REVIEW

**Comment by joeyparrish** [comment]
Just this (feature) PR and one more to merge before we all leave for the holidays.

File: shaka-lab-node/macos/shaka-lab-node.rb [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+  version "1.0.0"
+  sha256 "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
+
+  # Use --with-java to have Homebrew install Java from source.
+  # Skip that if you already have Oracle Java installed.
+  depends_on "java" => :optional
+
+  # Use --with-docker to have Homebrew install Docker from source.
+  # Docker is only needed to activate the Tizen node.
+  # Skip that if you already have Docker installed or don't need TIzen.
```
Comment (theodab) [inline shaka-lab-node/macos/shaka-lab-node.rb]: Nit: "TIzen" -> "Tizen"

File: shaka-lab-node/macos/log-wrapper.js [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+// Wrap macOS services for logging purposes.
+//
+// Services based on launchd for macOS can have their output redirected to a
+// log file.  However, this does not rotate logs.  External log rotation
+// mechanisms (such as newsyslog) want to signal the log writer to reopen the
+// log files on rotation, but launchd doesn't support this signal.
+// Therefore, the only way to get service output logging with rotation on macos
+// is to have the service itself write the log files and handle signals.
+//
+// This wrapper allows the service (and its child processes) continue logging
```
Comment (theodab) [inline shaka-lab-node/macos/log-wrapper.js]: Nit: It allows the service TO continue logging to etc etc.

File: shaka-lab-node/macos/README.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+TODO: Package distribution
+
+## Updates
+
+TODO: Package distribution
+
+## Configuration
+
+The config file is at `/opt/homebrew/etc/shaka-lab-node-config.yaml`.
+See the [configuration section](#configuration) below for details.
```
Comment (theodab) [inline shaka-lab-node/macos/README.md]: You know, I hadn't noticed this on read-throughs of the previous PRs, but why does the configuration section contain a hyperlink that links to... itself? And why does it say that the configuration section is "below", when it's referring to itself?

**Review by theodab — CHANGES_REQUESTED** [review]

File: shaka-lab-node/macos/shaka-lab-node.rb (hunk 1 above)
Comment (joeyparrish) [reply shaka-lab-node/macos/shaka-lab-node.rb]: Fixed

File: shaka-lab-node/macos/log-wrapper.js (hunk 2 above)
Comment (joeyparrish) [reply shaka-lab-node/macos/log-wrapper.js]: Fixed

File: shaka-lab-node/macos/README.md (hunk 3 above)
Comment (joeyparrish) [reply shaka-lab-node/macos/README.md]: Ah, I used to have all of the platform details in one giant doc, with a common configuration section.  So this is a refactoring mistake.  The config section is now in the doc one level up.  So both the link and text are wrong.  I'll fix all three in this PR.

Comment (joeyparrish) [reply shaka-lab-node/macos/README.md]: Fixed.

File: shaka-lab-node/macos/log-wrapper.js [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+  const args = process.argv.slice(2);
+  if (args.length < 4) {
+    console.log('Usage: node log-wrapper.js' +
+                ' <STDOUT-LOG> <STDERR-LOG> <PID-FILE> <COMMAND> [...]');
+    console.log('Pass a blank string as a log path to ignore that stream.');
+    process.exit(1);
+  }
+
+  // Extract arguments.
+  const stdoutLogPath = args[0];
```
Comment (alekzz) [inline shaka-lab-node/macos/log-wrapper.js]: nit; alternatively, an array destructuring can be used:

`const [stdoutLogPath, stderrLogPath, pidPath, command, ...commandArgs] = args;`

File: shaka-lab-node/macos/log-wrapper.js [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+      return fs.openSync(path, 'a');
+    } catch (error) {
+      console.log('Failed to open file:', path);
+      console.log(error);
+      process.exit(1);
+    }
+  };
+
+  const closeFd = (fd) => {
+    if (fd != null) {
```
Comment (alekzz) [inline shaka-lab-node/macos/log-wrapper.js:68]: nit;

May be better to use strict equality everywhere !==. I don't know about the style convention in this project though.

**Review by alekzz — DISMISSED** [review]

**Review by theodab — DISMISSED** [review]

File: shaka-lab-node/macos/log-wrapper.js (hunk 5 above)
Comment (joeyparrish) [reply shaka-lab-node/macos/log-wrapper.js:68]: The only thing that's == null but not === null is undefined.  So I don't think it makes a practical difference.

File: shaka-lab-node/macos/log-wrapper.js (hunk 4 above)
Comment (joeyparrish) [reply shaka-lab-node/macos/log-wrapper.js]: Thanks, that's a good point.  Done.

**Comment by joeyparrish** [comment]
Approvals were dismissed by the change I made to address Alex's feedback.  (Not as nice as Critique in that way.)  PTAL!

**Review by theodab — APPROVED** [review]

**Pull request merged**